Text/Megaparsec/Error.hs

```diff
@@ -248,6 +252,98 @@ parseErrorPretty
 parseErrorPretty e =
   sourcePosStackPretty (errorPos e) <> ":\n" <> parseErrorTextPretty e
 
+
+-- | Pretty-print a 'ParseError' and display the line on which the parse error
+-- occurred. The rendered 'String' always ends with a newline.
+--
+-- Assumes that the stream tokens can be coerced to an @Int@ and that the value
+-- integer value @10@ corresponds to a new-line character.
+--
+-- @since 6.0.0
+
+parseErrorPrettyWithLine
+  :: ( Enum (Token s)
+     , Ord t
+     , ShowToken t
+     , ShowErrorComponent e
+     , Stream s )
+  => s              -- ^ Original input stream
+  -> ParseError t e -- ^ Parse error to render
+  -> String         -- ^ Result of rendering
+parseErrorPrettyWithLine stream e = mconcat messageComponents
+  where
+    position = errorPos e
+    lineMessage = renderLineMessage stream $ NE.last position
+    messageComponents =
+      [ sourcePosStackPretty position
+      , ":\n"
+      , lineMessage 
+      , parseErrorTextPretty e ] 
+
+-- | Renders the source context to improve error message quality.
+
+renderLineMessage
+  :: forall s .
+     ( Enum (Token s)
+     , Stream s )
+  => s -> SourcePos -> String
+renderLineMessage s p = unlines [ paddingLine, contextLine, pointingLine ]
+  where
+    paddingLine   = padding <> "|"
+    contextLine   = " " <> lineNumberStr <> " | " <> cs
+    pointingLine  = padding <> "|" <> replicate columnNum ' ' <> "^"
+
+    lineNumberStr = show $ unPos linePos
+    padLength = length lineNumberStr + 2
+    padding = replicate padLength ' '
+
+    cs =
+      case ts of
+        [] -> "<empty line>"
+        xs -> chr . fromEnum <$> xs
+    ts = chunkToTokens (Proxy :: Proxy s) lineVal
+    lineVal = streamIter grabLine linePos s
```

Comment:
I see `streamIter` and `grabLine` are only used once here, for dropping `linePos` lines from the beginning of the stream. I would like you to merge the both functions into a single function that just clearly does what we need here: dropping N lines and returning the line we arrive at after that. This will be more clear to future readers of the code.